File: apps/ios/Shared/Views/UserSettings/NetworkAndServers/ChatRelayView.swift
//
//  ChatRelayView.swift
//  SimpleX (iOS)
//
//  Created by spaced4ndy on 23.02.2026.
//  Copyright © 2026 SimpleX Chat. All rights reserved.
//
// Spec: spec/architecture.md

import SwiftUI
import SimpleXChat

@ViewBuilder func showRelayTestStatus(relay: UserChatRelay) -> some View {
    switch relay.tested {
    case .some(true): Image(systemName: "checkmark").foregroundColor(.green)
    case .some(false): Image(systemName: "multiply").foregroundColor(.red)
    case .none: Color.clear
    }
}

func validRelayName(_ name: String) -> Bool {
    name != "" && validDisplayName(name)
}

func showInvalidRelayNameAlert(_ name: Binding<String>) {
    let validName = mkValidName(name.wrappedValue)
    if validName == "" {
        showAlert(NSLocalizedString("Invalid name!", comment: "alert title"))
    } else {
        showAlert(
            NSLocalizedString("Invalid name!", comment: "alert title"),
            message: String.localizedStringWithFormat(NSLocalizedString("Correct name to %@?", comment: "alert message"), validName),
            actions: {[
                UIAlertAction(title: NSLocalizedString("Ok", comment: "alert action"), style: .default) { _ in
                    name.wrappedValue = validName
                },
                cancelAlertAction
            ]}
        )
    }
}

func validRelayAddress(_ address: String) -> Bool {
    if let parsedMd = parseSimpleXMarkdown(address),
       parsedMd.count == 1,
       case .simplexLink(_, .relay, _, _) = parsedMd.first?.format {
        true
    } else {
        false
    }
}

func addChatRelay(
    _ relay: UserChatRelay,
    _ userServers: Binding<[UserOperatorServers]>,
    _ serverErrors: Binding<[UserServersError]>,
    _ serverWarnings: Binding<[UserServersWarning]>? = nil,
    _ dismiss: DismissAction
) {
    let nameEmpty = relay.name.trimmingCharacters(in: .whitespaces).isEmpty
    let addressEmpty = relay.address.trimmingCharacters(in: .whitespaces).isEmpty
    if nameEmpty && addressEmpty {
        dismiss()
    } else if !validRelayName(relay.name) {
        dismiss()
        showAlert(
            NSLocalizedString("Invalid relay name!", comment: "alert title"),
            message: NSLocalizedString("Check relay name and try again.", comment: "alert message")
        )
    } else if !validRelayAddress(relay.address) {
        dismiss()
        showAlert(
            NSLocalizedString("Invalid relay address!", comment: "alert title"),
            message: NSLocalizedString("Check relay address and try again.", comment: "alert message")
        )
    } else if let i = userServers.wrappedValue.firstIndex(where: { $0.operator == nil }) {
        userServers[i].wrappedValue.chatRelays.append(relay)
        validateServers_(userServers, serverErrors, serverWarnings)
        dismiss()
    } else { // Shouldn't happen
        dismiss()
        showAlert(NSLocalizedString("Error adding relay", comment: "alert title"))
    }
}

struct ChatRelayView: View {
    @Environment(\.dismiss) var dismiss: DismissAction
    @EnvironmentObject var theme: AppTheme
    @Binding var userServers: [UserOperatorServers]
    @Binding var serverErrors: [UserServersError]
    @Binding var serverWarnings: [UserServersWarning]
    @Binding var relay: UserChatRelay
    @State var relayToEdit: UserChatRelay
    var backLabel: LocalizedStringKey
    @State private var showTestFailure = false
    @State private var testing = false
    @State private var testFailure: RelayTestFailure?

    var body: some View {
        let validName = validRelayName(relayToEdit.name)
        let validAddress = validRelayAddress(relayToEdit.address)
        ZStack {
            if relay.preset {
                presetRelay()
            } else {
                customRelay(validName: validName, validAddress: validAddress)
            }
            if testing {
                ProgressView().scaleEffect(2)
            }
        }
        .modifier(BackButton(label: backLabel, disabled: Binding.constant(false)) {
            if validName && validAddress {
                relay = relayToEdit
                validateServers_($userServers, $serverErrors, $serverWarnings)
                dismiss()
            } else if !validName {
                dismiss()
                showAlert(
                    NSLocalizedString("Invalid relay name!", comment: "alert title"),
                    message: NSLocalizedString("Check relay name and try again.", comment: "alert message")
                )
            } else {
                dismiss()
                showAlert(
                    NSLocalizedString("Invalid relay address!", comment: "alert title"),
                    message: NSLocalizedString("Check relay address and try again.", comment: "alert message")
                )
            }
        })
        .alert(isPresented: $showTestFailure) {
            Alert(
                title: Text("Relay test failed!"),
                message: Text(testFailure?.localizedDescription ?? "")
            )
        }
        .onChange(of: relayToEdit.address) { _ in
            relayToEdit.tested = relayToEdit.address == relay.address ? relay.tested : nil
        }
    }

    private func relayNameHeader(validName: Bool) -> some View {
        HStack {
            Text("Your relay name").foregroundColor(theme.colors.secondary)
            if !validName {
                Spacer()
                Image(systemName: "exclamationmark.circle").foregroundColor(.red)
                    .onTapGesture { showInvalidRelayNameAlert($relayToEdit.name) }
            }
        }
    }

    private func presetRelay() -> some View {
        List {
            Section(header: Text("Preset relay name").foregroundColor(theme.colors.secondary)) {
                Text(relayToEdit.name)
            }
            Section(header: Text("Preset relay address").foregroundColor(theme.colors.secondary)) {
                Text(relayToEdit.address)
                    .textSelection(.enabled)
            }
            useRelaySection()
        }
    }

    private func customRelay(validName: Bool, validAddress: Bool) -> some View {
        List {
            Section {
                TextField("Enter relay name…", text: $relayToEdit.name)
                    .autocorrectionDisabled(true)
            } header: {
                relayNameHeader(validName: validName)
            }
            Section {
                TextEditor(text: $relayToEdit.address)
                    .multilineTextAlignment(.leading)
                    .autocorrectionDisabled(true)
                    .autocapitalization(.none)
                    .allowsTightening(true)
                    .lineLimit(10)
                    .frame(height: 144)
                    .padding(-6)
            } header: {
                HStack {
                    Text("Your relay address")
                        .foregroundColor(theme.colors.secondary)
                    if !validAddress {
                        Spacer()
                        Image(systemName: "exclamationmark.circle").foregroundColor(.red)
                    }
                }
            }
            useRelaySection(valid: validAddress)
            Section {
                Button(role: .destructive) {
                    relay.deleted = true
                    validateServers_($userServers, $serverErrors, $serverWarnings)
                    dismiss()
                } label: {
                    Label("Delete relay", systemImage: "trash")
                        .foregroundColor(.red)
                }
            }
        }
    }

    private func useRelaySection(valid: Bool = true) -> some View {
        Section(header: Text("Use relay").foregroundColor(theme.colors.secondary)) {
            HStack {
                Button("Test relay") {
                    testing = true
                    relayToEdit.tested = nil
                    Task {
                        if let f = await testRelayConnection(relay: $relayToEdit) {
                            showTestFailure = true
                            testFailure = f
                        }
                        await MainActor.run { testing = false }
                    }
                }
                .disabled(!valid || testing)
                Spacer()
                showRelayTestStatus(relay: relayToEdit)
            }
            Toggle("Use for new channels", isOn: $relayToEdit.enabled)
        }
    }
}

struct ChatRelayViewLink: View {
    @EnvironmentObject var theme: AppTheme
    @Binding var userServers: [UserOperatorServers]
    @Binding var serverErrors: [UserServersError]
    @Binding var serverWarnings: [UserServersWarning]
    @Binding var relay: UserChatRelay
    var duplicateRelayNames: Set<String>
    var duplicateRelayAddresses: Set<String>
    var backLabel: LocalizedStringKey
    @Binding var selectedServer: String?

    var body: some View {
        NavigationLink(tag: relay.id, selection: $selectedServer) {
            ChatRelayView(
                userServers: $userServers,
                serverErrors: $serverErrors,
                serverWarnings: $serverWarnings,
                relay: $relay,
                relayToEdit: relay,
                backLabel: backLabel
            )
            .navigationBarTitle("Chat relay")
            .modifier(ThemedBackground(grouped: true))
            .navigationBarTitleDisplayMode(.large)
        } label: {
            HStack {
                Group {
                    if duplicateRelayNames.contains(relay.name) || duplicateRelayAddresses.contains(relay.address) {
                        Image(systemName: "exclamationmark.circle").foregroundColor(.red)
                    } else if !relay.enabled {
                        Image(systemName: "slash.circle").foregroundColor(theme.colors.secondary)
                    } else {
                        showRelayTestStatus(relay: relay)
                    }
                }
                .frame(width: 16, alignment: .center)
                .padding(.trailing, 4)

                let displayName = !relay.name.isEmpty ? relay.name : relay.domains.first ?? relay.address
                let v = Text(displayName).lineLimit(1)
                if relay.enabled {
                    v
                } else {
                    v.foregroundColor(theme.colors.secondary)
                }
            }
        }
    }
}

struct NewChatRelayView: View {
    @Environment(\.dismiss) var dismiss: DismissAction
    @EnvironmentObject var theme: AppTheme
    @Binding var userServers: [UserOperatorServers]
    @Binding var serverErrors: [UserServersError]
    @Binding var serverWarnings: [UserServersWarning]
    @State private var relayToEdit = UserChatRelay(
        chatRelayId: nil, address: "", name: "", domains: [],
        preset: false, tested: nil, enabled: true, deleted: false
    )
    @State private var showTestFailure = false
    @State private var testing = false
    @State private var testFailure: RelayTestFailure?

    var body: some View {
        let validName = validRelayName(relayToEdit.name)
        let validAddress = validRelayAddress(relayToEdit.address)
        ZStack {
        List {
            Section {
                TextField("Enter relay name…", text: $relayToEdit.name)
                    .autocorrectionDisabled(true)
            } header: {
                HStack {
                    Text("Your relay name").foregroundColor(theme.colors.secondary)
                    if !validName {
                        Spacer()
                        Image(systemName: "exclamationmark.circle").foregroundColor(.red)
                            .onTapGesture { showInvalidRelayNameAlert($relayToEdit.name) }
                    }
                }
            }
            Section {
                TextEditor(text: $relayToEdit.address)
                    .multilineTextAlignment(.leading)
                    .autocorrectionDisabled(true)
                    .autocapitalization(.none)
                    .allowsTightening(true)
                    .lineLimit(10)
                    .frame(height: 144)
                    .padding(-6)
            } header: {
                HStack {
                    Text("Your relay address")
                        .foregroundColor(theme.colors.secondary)
                    if !validAddress {
                        Spacer()
                        Image(systemName: "exclamationmark.circle").foregroundColor(.red)
                    }
                }
            }
            Section(header: Text("Use relay").foregroundColor(theme.colors.secondary)) {
                HStack {
                    Button("Test relay") {
                        testing = true
                        relayToEdit.tested = nil
                        Task {
                            if let f = await testRelayConnection(relay: $relayToEdit) {
                                showTestFailure = true
                                testFailure = f
                            }
                            await MainActor.run { testing = false }
                        }
                    }
                    .disabled(!validAddress || testing)
                    Spacer()
                    showRelayTestStatus(relay: relayToEdit)
                }
                Toggle("Use for new channels", isOn: $relayToEdit.enabled)
            }
        }
        if testing {
            ProgressView().scaleEffect(2)
        }
        }
        .modifier(BackButton(disabled: Binding.constant(false)) {
            addChatRelay(relayToEdit, $userServers, $serverErrors, $serverWarnings, dismiss)
        })
        .alert(isPresented: $showTestFailure) {
            Alert(
                title: Text("Relay test failed!"),
                message: Text(testFailure?.localizedDescription ?? "")
            )
        }
    }
}

func testRelayConnection(relay: Binding<UserChatRelay>) async -> RelayTestFailure? {
    do {
        let (relayProfile, testFailure) = try await testChatRelay(address: relay.wrappedValue.address)
        if let f = testFailure {
            await MainActor.run { relay.wrappedValue.tested = false }
            return f
        }
        await MainActor.run {
            relay.wrappedValue.tested = true
            if let relayProfile {
                relay.wrappedValue.name = relayProfile.name
            }
        }
        return nil
    } catch {
        logger.error("testRelayConnection \(responseError(error))")
        await MainActor.run { relay.wrappedValue.tested = false }
        return nil
    }
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/usersettings/networkAndServers/ChatRelayView.kt
package chat.simplex.common.views.usersettings.networkAndServers

import SectionBottomSpacer
import SectionDividerSpaced
import SectionItemView
import SectionItemViewSpaceBetween
import SectionTextFooter
import SectionView
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.sp
import androidx.compose.ui.graphics.Color
import dev.icerock.moko.resources.compose.painterResource
import dev.icerock.moko.resources.compose.stringResource
import androidx.compose.ui.unit.dp
import chat.simplex.common.model.*
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.*
import chat.simplex.common.views.helpers.*
import chat.simplex.common.views.usersettings.PreferenceToggle
import chat.simplex.res.MR
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch

@Composable
fun ShowRelayTestStatus(relay: UserChatRelay, modifier: Modifier = Modifier) =
  when (relay.tested) {
    true -> Icon(painterResource(MR.images.ic_check), null, modifier, tint = SimplexGreen)
    false -> Icon(painterResource(MR.images.ic_close), null, modifier, tint = MaterialTheme.colors.error)
    else -> Icon(painterResource(MR.images.ic_check), null, modifier, tint = Color.Transparent)
  }

fun validRelayName(name: String): Boolean =
  name.isNotEmpty() && isValidDisplayName(name)

fun showInvalidRelayNameAlert(name: MutableState<String>) {
  val validName = mkValidName(name.value)
  if (validName.isEmpty()) {
    AlertManager.shared.showAlertMsg(
      title = generalGetString(MR.strings.invalid_name)
    )
  } else {
    AlertManager.shared.showAlertDialog(
      title = generalGetString(MR.strings.invalid_name),
      text = String.format(generalGetString(MR.strings.correct_name_to), validName),
      onConfirm = {
        name.value = validName
      }
    )
  }
}

fun validRelayAddress(address: String): Boolean {
  val parsedMd = parseToMarkdown(address)
  return parsedMd != null &&
    parsedMd.size == 1 &&
    parsedMd.first().format is Format.SimplexLink &&
    (parsedMd.first().format as Format.SimplexLink).linkType == SimplexLinkType.relay
}

fun addChatRelay(
  relay: UserChatRelay,
  userServers: MutableState<List<UserOperatorServers>>,
  serverErrors: MutableState<List<UserServersError>>,
  serverWarnings: MutableState<List<UserServersWarning>>?,
  rhId: Long?,
  close: () -> Unit
) {
  val nameEmpty = relay.name.trim().isEmpty()
  val addressEmpty = relay.address.trim().isEmpty()
  if (nameEmpty && addressEmpty) {
    close()
  } else if (!validRelayName(relay.name)) {
    close()
    AlertManager.shared.showAlertMsg(
      title = generalGetString(MR.strings.invalid_relay_name),
      text = generalGetString(MR.strings.check_relay_name)
    )
  } else if (!validRelayAddress(relay.address)) {
    close()
    AlertManager.shared.showAlertMsg(
      title = generalGetString(MR.strings.invalid_relay_address),
      text = generalGetString(MR.strings.check_relay_address)
    )
  } else {
    val i = userServers.value.indexOfFirst { it.operator == null }
    if (i != -1) {
      val updatedUserServers = userServers.value.toMutableList()
      val operatorServers = updatedUserServers[i]
      updatedUserServers[i] = operatorServers.copy(
        chatRelays = operatorServers.chatRelays + relay
      )
      userServers.value = updatedUserServers
      withBGApi {
        validateServers_(rhId, userServers.value, serverErrors, serverWarnings)
      }
      close()
    } else { // Shouldn't happen
      close()
      AlertManager.shared.showAlertMsg(title = generalGetString(MR.strings.error_adding_relay))
    }
  }
}

@Composable
fun ChatRelayView(
  relay: UserChatRelay,
  onDelete: () -> Unit,
  onUpdate: (UserChatRelay) -> Unit,
  close: () -> Unit
) {
  val relayToEdit = remember { mutableStateOf(relay) }

  LaunchedEffect(Unit) {
    snapshotFlow { relayToEdit.value.address }
      .distinctUntilChanged()
      .collect {
        if (relayToEdit.value.address == relay.address) {
          relayToEdit.value = relayToEdit.value.copy(tested = relay.tested, relayProfile = relay.relayProfile)
        } else {
          relayToEdit.value = relayToEdit.value.copy(tested = null)
        }
      }
  }

  ModalView(
    close = {
      val validName = validRelayName(relayToEdit.value.name)
      val validAddress = validRelayAddress(relayToEdit.value.address)
      if (validName && validAddress) {
        onUpdate(relayToEdit.value)
        close()
      } else if (!validName) {
        close()
        AlertManager.shared.showAlertMsg(
          title = generalGetString(MR.strings.invalid_relay_name),
          text = generalGetString(MR.strings.check_relay_name)
        )
      } else {
        close()
        AlertManager.shared.showAlertMsg(
          title = generalGetString(MR.strings.invalid_relay_address),
          text = generalGetString(MR.strings.check_relay_address)
        )
      }
    }
  ) {
    ChatRelayLayout(
      relayToEdit,
      onDelete = onDelete
    )
  }
}

@Composable
private fun ChatRelayLayout(
  relay: MutableState<UserChatRelay>,
  onDelete: (() -> Unit)?
) {
  val testing = remember { mutableStateOf(false) }
  Box {
    ColumnWithScrollBar {
      AppBarTitle(stringResource(MR.strings.chat_relay))
      if (relay.value.preset) {
        PresetRelay(relay, testing)
      } else {
        CustomRelay(relay, onDelete, testing)
      }
      SectionBottomSpacer()
    }
    if (testing.value) {
      DefaultProgressView(null)
    }
  }
}

@Composable
private fun PresetRelay(relay: MutableState<UserChatRelay>, testing: MutableState<Boolean>) {
  SectionView(stringResource(MR.strings.preset_relay_address).uppercase()) {
    SelectionContainer {
      Text(
        relay.value.address,
        Modifier.padding(start = DEFAULT_PADDING, top = 5.dp, end = DEFAULT_PADDING, bottom = 10.dp),
        color = MaterialTheme.colors.secondary
      )
    }
  }
  SectionDividerSpaced()
  SectionView(stringResource(MR.strings.preset_relay_name).uppercase()) {
    SectionItemView {
      Text(relay.value.name)
    }
  }
  SectionDividerSpaced()
  UseRelaySection(relay, testing = testing)
}

@Composable
private fun CustomRelay(
  relay: MutableState<UserChatRelay>,
  onDelete: (() -> Unit)?,
  testing: MutableState<Boolean>
) {
  val relayName = remember { mutableStateOf(relay.value.name) }
  val relayAddress = remember { mutableStateOf(relay.value.address) }
  val validName = remember { derivedStateOf { validRelayName(relayName.value) } }
  val validAddress = remember { derivedStateOf { validRelayAddress(relayAddress.value) } }

  LaunchedEffect(Unit) {
    snapshotFlow { relayName.value }
      .distinctUntilChanged()
      .collect { relay.value = relay.value.copyWithName(it) }
  }
  LaunchedEffect(Unit) {
    snapshotFlow { relay.value.name }
      .distinctUntilChanged()
      .collect { relayName.value = it }
  }
  LaunchedEffect(Unit) {
    snapshotFlow { relayAddress.value }
      .distinctUntilChanged()
      .collect { relay.value = relay.value.copy(address = it) }
  }

  SectionView(
    stringResource(MR.strings.your_relay_address).uppercase(),
    icon = painterResource(MR.images.ic_error),
    iconTint = if (!validAddress.value) MaterialTheme.colors.error else Color.Transparent,
  ) {
    TextEditor(
      relayAddress,
      Modifier.height(144.dp)
    )
  }
  SectionDividerSpaced(maxTopPadding = true)

  Column {
    val iconSize = with(LocalDensity.current) { 21.sp.toDp() }
    Row(Modifier.padding(start = DEFAULT_PADDING, bottom = 5.dp), verticalAlignment = Alignment.CenterVertically) {
      Text(
        stringResource(MR.strings.your_relay_name).uppercase(),
        color = MaterialTheme.colors.secondary, style = MaterialTheme.typography.body2, fontSize = 12.sp
      )
      IconButton(
        onClick = { if (!validName.value) showInvalidRelayNameAlert(relayName) },
        enabled = !validName.value,
        modifier = Modifier.padding(start = DEFAULT_PADDING_HALF).size(iconSize)
      ) {
        Icon(
          painterResource(MR.images.ic_error), null,
          tint = if (!validName.value) MaterialTheme.colors.error else Color.Transparent
        )
      }
    }
    Column(Modifier.fillMaxWidth()) {
      TextEditor(
        relayName,
        Modifier,
        placeholder = generalGetString(MR.strings.enter_relay_name),
        enabled = relay.value.tested != true
      )
    }
  }
  if (relay.value.tested != true) {
    SectionTextFooter(annotatedStringResource(MR.strings.test_relay_to_retrieve_name))
  }
  SectionDividerSpaced(maxTopPadding = true)

  UseRelaySection(relay, validAddress.value, testing)

  if (onDelete != null) {
    SectionDividerSpaced()
    SectionView {
      SectionItemView(onDelete) {
        Text(stringResource(MR.strings.delete_relay), color = MaterialTheme.colors.error)
      }
    }
  }
}

@Composable
private fun UseRelaySection(
  relay: MutableState<UserChatRelay>,
  valid: Boolean = true,
  testing: MutableState<Boolean>
) {
  val scope = rememberCoroutineScope()
  SectionView(stringResource(MR.strings.use_relay).uppercase()) {
    SectionItemViewSpaceBetween(
      click = {
        testing.value = true
        relay.value = relay.value.copy(tested = null)
        scope.launch {
          val f = testRelayConnection(relay)
          if (f != null) {
            AlertManager.shared.showAlertMsg(
              title = generalGetString(MR.strings.relay_test_failed_alert),
              text = f.localizedDescription
            )
          }
          testing.value = false
        }
      },
      disabled = !valid || testing.value
    ) {
      Text(
        stringResource(MR.strings.test_relay),
        color = if (valid && !testing.value) MaterialTheme.colors.onBackground else MaterialTheme.colors.secondary
      )
      ShowRelayTestStatus(relay.value)
    }

    val enabled = rememberUpdatedState(relay.value.enabled)
    PreferenceToggle(
      stringResource(MR.strings.use_for_new_channels),
      checked = enabled.value
    ) {
      relay.value = relay.value.copy(enabled = it)
    }
  }
}

@Composable
fun ChatRelayViewLink(
  relay: UserChatRelay,
  duplicateRelayNames: Set<String>,
  duplicateRelayAddresses: Set<String>,
  onClick: () -> Unit
) {
  SectionItemView(onClick) {
    Box(Modifier.width(16.dp)) {
      when {
        relay.name in duplicateRelayNames || relay.address in duplicateRelayAddresses -> InvalidServer()
        !relay.enabled -> Icon(painterResource(MR.images.ic_do_not_disturb_on), null, tint = MaterialTheme.colors.secondary)
        else -> ShowRelayTestStatus(relay)
      }
    }
    Spacer(Modifier.padding(horizontal = 4.dp))
    val displayName = relay.name.ifEmpty { relay.domains.firstOrNull() ?: relay.address }
    if (relay.enabled) {
      Text(displayName, color = MaterialTheme.colors.onBackground, maxLines = 1)
    } else {
      Text(displayName, maxLines = 1, color = MaterialTheme.colors.secondary)
    }
  }
}

@Composable
fun ModalData.NewChatRelayView(
  userServers: MutableState<List<UserOperatorServers>>,
  serverErrors: MutableState<List<UserServersError>>,
  serverWarnings: MutableState<List<UserServersWarning>>,
  rhId: Long?,
  close: () -> Unit
) {
  val relayToEdit = remember {
    mutableStateOf(
      UserChatRelay(
        chatRelayId = null, address = "", name = "", domains = emptyList(),
        preset = false, tested = null, enabled = true, deleted = false
      )
    )
  }

  LaunchedEffect(Unit) {
    snapshotFlow { relayToEdit.value.address }
      .distinctUntilChanged()
      .collect {
        relayToEdit.value = relayToEdit.value.copy(tested = null)
      }
  }

  ModalView(close = {
    addChatRelay(relayToEdit.value, userServers, serverErrors, serverWarnings, rhId, close)
  }) {
    NewChatRelayLayout(relayToEdit)
  }
}

@Composable
private fun NewChatRelayLayout(relay: MutableState<UserChatRelay>) {
  val testing = remember { mutableStateOf(false) }
  Box {
    ColumnWithScrollBar {
      AppBarTitle(stringResource(MR.strings.new_chat_relay))
      CustomRelay(relay, onDelete = null, testing = testing)
      SectionBottomSpacer()
    }
    if (testing.value) {
      DefaultProgressView(null)
    }
  }
}

suspend fun testRelayConnection(relay: MutableState<UserChatRelay>): RelayTestFailure? =
  try {
    val (relayProfile, testFailure) = chatModel.controller.testChatRelay(chatModel.remoteHostId(), relay.value.address)
    if (testFailure != null) {
      relay.value = relay.value.copy(tested = false)
      testFailure
    } else {
      relay.value = relay.value.copy(tested = true).let {
        if (relayProfile != null) it.copyWithName(relayProfile.name) else it
      }
      null
    }
  } catch (e: Exception) {
    Log.e(TAG, "testRelayConnection ${e.stackTraceToString()}")
    relay.value = relay.value.copy(tested = false)
    null
  }
